### PR TITLE
Add incremental build and project references to java projects

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: eng/targets/Java.Common.props [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <PropertyGroup>
+    <!-- Disable gradle daemon on CI since the CI seems to try to wait for the daemon to shut down, which it doesn't do -->
+    <GradleOptions Condition="'$(ContinuousIntegrationBuild)' == 'true'">$(GradleOptions) -Dorg.gradle.daemon=false</GradleOptions>
+    <Configuration Condition="'$(Configuration)' == '' AND '$(ContinuousIntegrationBuild)' == 'true'">Release</Configuration>
+    <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
```
Comment (dougbu) [inline eng/targets/Java.Common.props:6]: Why is this needed❔ `$(Configuration)` should always be set at least when using our `build.*` scripts.

File: eng/targets/Java.Common.targets [hunk 2]
```diff
@@ -0,0 +1,87 @@
+<Project DefaultTargets="Build">
+
+  <PropertyGroup>
+    <BaseIntermediateOutputPath Condition="'$(BaseIntermediateOutputPath)' == ''">$(MSBuildProjectDirectory)\obj\</BaseIntermediateOutputPath>
+    <IntermediateOutputPath>$([MSBuild]::NormalizeDirectory('$(BaseIntermediateOutputPath)'))$(Configuration)\</IntermediateOutputPath>
```
Comment (dougbu) [inline eng/targets/Java.Common.targets:5]: Could we get the Java builds using the regular artifacts/bin and artifacts/obj location❔ All exceptions from that convention make `rm artifacts -recurse` useless.

File: eng/targets/Java.Common.targets [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      ResolveProjectReferences;
+      _Build;
+    </BuildDependsOn>
+    <JavaBuildArgs Condition="'$(JavaBuildArgs)' == ''">../gradlew $(GradleOptions) compileJava</JavaBuildArgs>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <JavaFiles Include="src\**\*.java" />
+    <JavaFiles Include="*.javaproj" />
+    <JavaFiles Include="build.gradle" />
```
Comment (dougbu) [inline eng/targets/Java.Common.targets:17]: Nice.

Curious: Why does a `*.javaproj` project need to include itself in this item group❔

File: eng/targets/Java.Common.targets [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    <BuildOutputFiles Include="$(BaseIntermediateOutputPath)build-sentinel" />
+  </ItemGroup>
+
+  <Target Name="Restore" />
+
+  <PropertyGroup>
+    <PackDependsOn Condition=" '$(NoBuild)' != 'true' ">
+      $(PackDependsOn);
+      Build
+    </PackDependsOn>
```
Comment (dougbu) [inline eng/targets/Java.Common.targets:28]: This setting is repeated below. Remove one of them. Also, is the order correct here i.e. should `Build` be first❔

File: eng/targets/Java.Common.targets [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  <PropertyGroup>
+    <PackDependsOn>
+      $(PackDependsOn);
+      PrepareForBuild
+    </PackDependsOn>
+    <PackDependsOn Condition="'$(NoBuild)' != 'true'">
+      $(PackDependsOn);
+      Build;
+    </PackDependsOn>
```
Comment (dougbu) [inline eng/targets/Java.Common.targets:72]: With the current settings, `$(PackDependsOn)` will contain at least `Build;PrepareForBuild;Build`. Suspect that means the earlier setting is the one that should be removed.

File: src/SignalR/clients/java/signalr/core/signalr.client.java.core.javaproj [hunk 6]
```diff
@@ -21,37 +19,10 @@
...
-  <Target Name="Restore" />
-
-  <PropertyGroup>
-    <PackDependsOn Condition=" '$(NoBuild)' != 'true' ">
-      $(PackDependsOn);
-      Build
-    </PackDependsOn>
-  </PropertyGroup>
-
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.targets))\Directory.Build.targets" />
```
Comment (dougbu) [inline src/SignalR/clients/java/signalr/core/signalr.client.java.core.javaproj:22]: nit: Can this be at the end of the file now, matching the usual conventions for projects that don't auto-import these files❔

File: src/SignalR/clients/java/signalr/messagepack/signalr.client.java.messagepack.javaproj [hunk 7]
```diff
@@ -21,37 +19,14 @@
...
-    <PackDependsOn Condition=" '$(NoBuild)' != 'true' ">
-      $(PackDependsOn);
-      Build
-    </PackDependsOn>
-  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="../core/signalr.client.java.core.javaproj" />
+  </ItemGroup>
 
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.targets))\Directory.Build.targets" />
```
Comment (dougbu) [inline src/SignalR/clients/java/signalr/messagepack/signalr.client.java.messagepack.javaproj:26]: Same nit as in src/SignalR/clients/java/signalr/core/signalr.client.java.core.javaproj

File: eng/targets/Java.Common.targets (hunk 3 above)
Comment (BrennanConroy) [reply eng/targets/Java.Common.targets:17]: If the javaproj is modified then build should be out-of-date. Don't think that happens unless it's added as part of the hash

File: eng/targets/Java.Common.targets [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    Outputs="@(BuildOutputFiles)">
+    <Telemetry EventName="NETCORE_ENGINEERING_TELEMETRY" EventData="Category=Build" />
+    <Exec Command="$(JavaBuildArgs)" />
+    <WriteLinesToFile Overwrite="true" File="$(BaseIntermediateOutputPath)build-sentinel" />
+  </Target>
+
+  <PropertyGroup>
+    <PackDependsOn>
+      $(PackDependsOn);
+      PrepareForBuild
```
Comment (halter73) [inline eng/targets/Java.Common.targets:67]: Do we still want to run PrepareForBuild if `'$(NoBuild)' == 'true'`?

**Review by wtgodbe — APPROVED** [review]
Thanks!

**Review by dougbu — APPROVED** [review]
